Define syntax for escaping declarative config env var substitution

jack-berg opened the pull request:
Resolves #3914. 

An environment variable substitution reference which starts with `$$` (i.e. `$${API_KEY}`) is escaped. Implementations should strip the leading `$` character and not perform substitution (i.e. `$${API_KEY} => ${API_KEY}`).  

See java implementation here: https://github.com/open-telemetry/opentelemetry-java/pull/7033

tsloughter commented:
Why `$` and not `\`? The latter being what I'd expect and try first when using this since it is what I'd do in a shell script.

Also doesn't this break being able to do something like `$${NUM_OF_DOLLARS}`?

tsloughter reviewed (approved):
Approving despite my comments since `$$` is what the collector uses and need for `$${DOLLARS}` seem unnecessary in an SDK configuration :)

marcalff reviewed (changes requested):
The escape sequence is not fully specified.

If I want `${API_KEY}` in the environment variable value, this says it should be encoded as `$${API_KEY}`, ok.

If I want `$${API_KEY}` in the value, how should this be encoded ?

If '$' is used to as an escape character, then the spec must have a well defined way to print a `$` non escaping character, so that any string containing `$` can be represented.

marcalff commented:
> The escape sequence is not fully specified.
> 
> If I want `${API_KEY}` in the environment variable value, this says it should be encoded as `$${API_KEY}`, ok.
> 
> If I want `$${API_KEY}` in the value, how should this be encoded ?
> 
> If '$' is used to as an escape character, then the spec must have a well defined way to print a `$` non escaped character, so that any string containing `$` can be represented.

Resolution could be to say that `$${API_KEY}` is encoded as `$$${API_KEY}`, and so forth, but this should be in the spec and not open to interpretation.

pellared commented:
Maybe we should do it simply in the same as `docker compose`. See https://docs.docker.com/reference/compose-file/interpolation/

> You can use a `$$` (double-dollar sign) when your configuration needs a literal dollar sign.

pellared reviewed (changes requested):
Maybe we should do it simply in the same as `docker compose`. See https://docs.docker.com/reference/compose-file/interpolation/

> You can use a `$$` (double-dollar sign) when your configuration needs a literal dollar sign.

Is also looks that it is what OTel Collector is doing:
- https://github.com/open-telemetry/opentelemetry-collector/blob/461a3558086a03ab13ea121d12e28e185a1c79b0/confmap/internal/e2e/testdata/expand-escaped-env.yaml

jack-berg commented:
> Maybe we should do it simply in the same way as docker compose. See https://docs.docker.com/reference/compose-file/interpolation/

>> You can use a $$ (double-dollar sign) when your configuration needs a literal dollar sign.

> Is also looks that it is what OTel Collector is doing as well:

This is the syntax proposed in this PR.

pellared commented:
> This is the syntax proposed in this PR.

I disagree. According to this PR something like `a $$ b` is evaluated to `a $$ b` whereas for Collector and Docker Compose it will be evaluated to `a $ b`. Meaning in both Collector and Docker Compose `$$` is always escaped to `$` and not only when it is matching the env var substitution regex.

jack-berg commented:
@tsloughter: 

> Approving despite my comments since $$ is what the collector uses

Yes exactly. Matching collector syntax is the motivation here.

---

@pellared:

> I disagree. According to this PR something like "a $$ b" is evaluated to a $$ b whereas for OTel Collector and Docker Compose it will be evaluated to a $ b. Meaning in both OTel Collector and Docker Compose $$ is always escaped to $ and not only when it is matching the env var substitution regex.

I see. That's fair.

When I was writing this PR I did look at the [collector's language](https://opentelemetry.io/docs/collector/configuration/#environment-variables), which is similar to docker's, and which I found to be overly simplistic:

> Use $$ to indicate a literal $.

Its missing guidance on how `$$` interacts with environment variable substitution - specifically, the ordering.

For example, suppose we have: `$${FOO}` and `FOO=bar`. 

If we perform env env vars substitution _after_ `$$` substitution, we get `$${FOO} => ${FOO} => bar`, which is not what the user wants.

Alternatively, env var substitution might take place _before_ `$$` substitution, we get `$${FOO}` => $bar`, which is not what the user wants.

So `$$` subsitution can't happen _before_ or _after_ env var substitution. Instead, it must happen simultaneously as a part of env var substitution. Hence, my inclusion of it in the env var regular expression.

You also gave an example of `"a $$ b"` evaluating to `"a $ b"`. What is the use of escaping `$` outside of references to env var substitution? A user who wishes to resolve `"a $ b"` can just enter that value as is - no escape needed.

---

@marcalff 

> If I want $${API_KEY} in the value, how should this be encoded ?
> Resolution could be to say that $${API_KEY} is encoded as $$${API_KEY}, and so forth, but this should be in the spec and not open to interpretation.

My understanding is that the regular expression / logic as proposed in this PR will indeed resolve `$$${API_KEY}` to be `$${API_KEY}`. 

The reason is that the regular expression `\${1,2}\{(?:env:)?(?<ENV_NAME>[a-zA-Z_][a-zA-Z0-9_]*)(:-(?<DEFAULT_VALUE>[^\n]*))?\}` matches exactly 1 OR 2 `$` characters. So for `$$${API_KEY}`, the first `$` is unmatched and left alone. The following characters, `$${API_KEY}`, matches the expression, but since it starts with `$$`, it is escaped instead with the first `$` dropped instead of replaced. So in total we have `$` + `${API_KEY}` = `$${API_KEY}`.

I added an [additional test case](https://github.com/open-telemetry/opentelemetry-java/pull/7033/commits/940dc1ab9757b9ca9fec2ffee5529d729dcb4d93) to my java prototype implementation of this. Note - no changes to the implementation were required to make this test pass.

More examples in the specification might be helpful.

tsloughter commented:
Put this comment in the wrong repo:

Oh, now I see the collector uses $$. Then I suppose this is the way to go. May want to also mention additional escaping like allowing \$${DOLLARS} to be $10?

pellared commented:
@jack-berg, I find the Collector and and Docker Compose docs clearer. `$${FOO}` is evaluated to `${FOO}`.

> You also gave an example of `"a $$ b"` evaluating to `"a $ b"`. What is the use of escaping `$` outside of references to env var substitution? A user who wishes to resolve `"a $ b"` can just enter that value as is - no escape needed.

This will provide different experience for users who have experience with Collector and Docker Compose who are used to always type `$$` when they want to add a `$` literal. For instance, `$$$${FOO}` should be evaluated to `$${FOO}`.

I agree with the intention to match with Collector (and Docker Compose) syntax  :+1:

jack-berg commented:
>  I find the Collector and and Docker Compose docs clearer. $${FOO} is evaluated to ${FOO}.

Its seems clear on paper until you try to implement it. The docker and collector specs are incomplete. 

Why does `$${FOO}` resolve to `${FOO}` and not `$${FOO} => $bar` (env var before `$$`) or `$${FOO} => ${FOO} => bar` (env var after `$$`)? 

If you try to implement this, and treat `$$` => `$` substitution separately from env var substitution, then you need to answer the question of ordering: which comes first? And as I noted above, both possible answers of ordering yield incorrect results.

I would prefer to match the algorithm of the collector, but looking for advice on how to describe this in such a way that 11 language implementations will get it right. Despite the seemingly simple syntax as described in docs, the [collector's substitution logic](https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/expand.go) is quite complex / nuanced. The current env var substitution allows for simpler implementations that can leverage regular expressions to do the heavy lifting.

I considered trying to get the both of both worlds (i.e. collector behavior and a regular expression based spec), by changing the regular expression to match:

```
\${1,2}(\{(?:env:)?(?<ENV_NAME>[a-zA-Z_][a-zA-Z0-9_]*)(:-(?<DEFAULT_VALUE>[^\n]*))?\})?
```

I.e. make the whole `{API_KEY:-default}` part optional so the expression still matches against just `$$`, but this still has ordering challenges. Consider @marcalff's example of wanting a resolved string of `$${API_KEY}`. How do you accomplish this? With the updated regular expression and assuming `API_KEY=value`, `$$${API_KEY}`, resolves to `$value`. This is because the expression matches twice: 

- the first `$$` matches and resolves to `$`
- the second `${API_KEY}` matches and resolves to value
 
I'm actually not sure there's a way to accomplish the collector's behavior without requiring implementations to carefully recreate its complex recursive algorithm which statefully walks values, character by character. (Not sure how I would approach trying to to specify this.) And I think a regular expression based approach is essential to having a low enough complexity bar to implement.

pellared commented:
Docker composer also does it recursively: https://github.com/compose-spec/compose-go/blob/main/interpolation/interpolation.go

> looking for advice on how to describe this in such a way that 11 language implementations will get it right

I think "specification by example" will help achieving it. Meaning specifying expected results for given input.

> regular expression based spec

Maybe this would help: https://github.com/compose-spec/compose-go/blob/main/template/template.go

> [collector's substitution logic](https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/expand.go) is quite complex / nuanced

> I'm actually not sure there's a way to accomplish the collector's behavior without requiring implementations to carefully recreate its complex recursive algorithm which statefully walks values, character by character.

As far as I remember, I was calling it out long ago during some SIG meeting that env var substitution is complex. I think that I even said that each language should have fuzz tests for such parsing because of its complexity. 

I was proposing that the users could simply use `envsubst` so that each SDK could make the parsing of the files a lot simpler and less bug-prone.

However, the feedback was that the Configuration SIG wants to try (challenge) it given the assumption that users need it.

If this won't be working like in OTel Collector then I think we will be giving bad user experience.

Given that OTel Collector is still not stable, the alternative way could be simplifying the env var substitution in both Collector config and SDK config.

pellared commented:
We could also provide a seperate OTel tool which does the env var substitution in the same way as the OTel Collector. This way we would do not have to implement it in each language, but just reuse the OTel Collector code. We could add env var interpolation as an opt-in feature in the SDKs (e.g. controled by an env var :wink:).

jack-berg commented:
> As far as I remember, I was calling it out long ago during some SIG meeting that env var substitution is complex.

It doesn't have to be complex. With a regex-based specification, the [java implementation](https://github.com/open-telemetry/opentelemetry-java/blob/13e2b8c4dfeba09f361b47986febbd792ab1ae60/sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java#L309-L337) is only 28 lines long, plus a little more boiler plate to make sure that logic is invoked by the YAML parser.

> We could also provide a seperate OTel tool which does the env var substitution in the same way as the OTel Collector.

We actually already have this. The [configuration validator](https://github.com/open-telemetry/opentelemetry-configuration/tree/main/validator) is a tool written in go, and available as a binary or docker container, which accepts a target YAML config file and:

- performs env var substiution
- validates the contents against the configuration JSON schema

The problem is, not all ecosystems will want or be able to bundle in a binary specify to a particular platform / architecture. So this tool is available to help implementations, but we cannot rely on it being used everywhere. 

> If this won't be working like in OTel Collector then I think we will be giving bad user experience.
> Given that OTel Collector is still not stable, the alternative way could be simplifying the env var substitution in both Collector config and SDK config.

As of this PR, the differences I know of are:
- collector substitutes `$$` for `$` even if not part of an env var substitution ref
- collector supports reference values from other [provides](https://github.com/open-telemetry/opentelemetry-collector/tree/4963da7b8eb3b1024bac9b3142ed695d7e97afee/confmap#provider) besides env vars, which configuration has punted on to reduce maintenance burden.

So there are already differences between collector and language implementations. Its probably safe to say that the env var substitution for declarative config is a subset of the collector. And this seems fine to me. Especially if its a strict subset (i.e. declarative config takes care not to introduce new things the collector will not add).

Not sure if @open-telemetry/collector-maintainers would be open to it, but it doesn't seem like much is lost by removing generic support for replacing `$$` with `$` outside of env var substitution references. WDYT?

I get the appeal "docker does this" argument, but docker env var substitution breaks from collector substitution and declarative configuration is other key ways. All try to have a syntax inspired by shell-style substitution, but each supports shell-style to varying extents, and the collector extends it with the provider concept which is distinctly different.

marcalff commented:
> So in total we have `$` + `${API_KEY}` = `$${API_KEY}`.

I see.

This is not trivial when reading the regexp that checks for 1 or 2 `$` characters, as in `\${1,2}`, but I agree it works as intended then.

Having examples so that every language implements the proper behavior is critical here.

pellared commented:
> The problem is, not all ecosystems will want or be able to bundle in a binary specify to a particular platform / architecture. So this tool is available to help implementations, but we cannot rely on it being used everywhere.

I created a separate issue: https://github.com/open-telemetry/opentelemetry-specification/issues/4384

I drifted away too much from the original issue and PR. Sorry for that :pray: (no bad intentions)

Regarding this PR I just expect the same env var substitution behavior from **all** OTel Components.

jack-berg commented:
Nice proposal @marcalff - I implemented this in my prototype PR in this [commit](https://github.com/open-telemetry/opentelemetry-java/pull/7033/commits/37d6bbbea5ee4ddc46795cd2f5271109a7f9cb1e). Its only slightly more complex that a pure regex based approach.

> The nasty part here is to convey sufficient information in the spec without having to produce a full parser grammar and reference litterature.

I agree, but I think your proposal balances it nicely. Let me take a shot at updating the text in this PR accordingly.

TylerHelmuth commented:
I agree with @marcalff's proposal at https://github.com/open-telemetry/opentelemetry-specification/pull/4375#issuecomment-2609387681 as this is the collector's current behavior.

See https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/internal/e2e/expand_test.go as a good list of examples for how the collector handles expansion/escaping. If you're wondering how the collector handles a specific scenario that is not already covered in those tests let me know and I'll add it.

jack-berg commented:
Ok I've pushed a commit to reflect @marcalff's [proposal](https://github.com/open-telemetry/opentelemetry-specification/pull/4375#issuecomment-2609387681), including some pseudocode of what an implementation might look like. I've also converted the examples section to a markdown table and added more examples per this conversation. The re-org improves readability as the set of examples grows.

Let me know what you think.

jack-berg commented on specification/configuration/data-model.md:
```diff
@@ -75,6 +75,29 @@ more non line break characters (i.e. any character except `\n`). If a referenced
 environment variable is not defined and does not have a `DEFAULT_VALUE`, it MUST
 be replaced with an empty value.
 
+The `$` character is an escape sequence, such that `$$` in the input is
```
TODO: add changelog entry before merging

marcalff commented on specification/configuration/data-model.md:
```diff
@@ -75,6 +75,29 @@ more non line break characters (i.e. any character except `\n`). If a referenced
 
+The `$` character is an escape sequence, such that `$$` in the input is
+translated to a single `$` in the output. The resolved `$` from an escape
+sequence MUST not be considered when matching input against the environment
```
MUST NOT ?

jack-berg replied:
I _think_ (correct me if I'm wrong) this is a critical part of the mechanics:

Assuming `API_KEY=123`.

If the resolved `$` is not considered when matching against environment variables, we have:

`"$${API_KEY}"` => `"$" + resolveEnvVars("{API_KEY}")` => `"${API_KEY}"`

If instead, the resolve `$` is considered when matching against environment variables, we could have something like:

`"$${API_KEY}"` => `"$" + "{API_KEY}"` => `resolveEnvVars("${API_KEY}")` => `"123"`.

In other words, without this, an parser might think that it can do the substitution in two separate passes. The first pass replacing `$$` with `$`. The second pass replacing env var substitution references. A two pass approach produces incorrect results.

pellared replied:
I guess Marc may simply had this in mind:

```suggestion
sequence MUST NOT be considered when matching input against the environment
```

marcalff replied:
Yes, sorry about the short comment. I meant to write MUST NOT in uppercase, per pellared change.

marcalff commented on specification/configuration/data-model.md:
```diff
@@ -94,55 +117,42 @@ example, see references to `INVALID_MAP_VALUE` environment variable below.
+| `key: $$${STRING_VALUE}`                  | `key: $value`               | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, `${STRING_VALUE}` is replaced with `value`                                                     |
+| `key: $$$${STRING_VALUE}`                 | `key: $${STRING_VALUE}`     | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, `$$` escape sequence is replaced with `$`, `{STRING_VALUE}` does not match substitution syntax |
+| `key: a $$ b`                             | `key: a $ b`                | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`                                                                                                 |
+| `key: a $ b`                              | `key: a $ b`                | `tag:yaml.org,2002:str`   | No escape sequence, no substitution references, value is left alone                                                                       |
```
Style,

"value is left alone" -> value is unchanged ?

marcalff reviewed (approved):
LGTM

mx-psi reviewed (commented):
Apologies for the late comment, but I think it would be useful to see an example where the escaping sequence is part of the environment variable value. The declarative configuration implementation will probably differ from the Collector's there, since it doesn't allow recursive expansion, so I think it is important to show it here

jack-berg commented on specification/configuration/data-model.md:
```diff
@@ -94,55 +117,48 @@ example, see references to `INVALID_MAP_VALUE` environment variable below.
+| `key: $${STRING_VALUE:-fallback}`          | `${STRING_VALUE:-fallback}`         | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, `{STRING_VALUE:-fallback}` does not match substitution syntax                                          |
+| `key: $${STRING_VALUE:-${STRING_VALUE}}`   | `${STRING_VALUE:-value}`            | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, leaving `{STRING_VALUE:-${STRING_VALUE}}`, `${STRING_VALUE}` is replaced with `value`                  |
+| `key: ${UNDEFINED_KEY:-$${UNDEFINED_KEY}}` | `${STRING_VALUE:-${UNDEFINED_KEY}}` | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, leaving `${UNDEFINED_KEY:-` before and `{UNDEFINED_KEY}}` after which do not match substitution syntax |
+| `key: ${VALUE_WITH_ESCAPE}`                | `value$$`                           | `tag:yaml.org,2002:str`   | Value of env var `VALUE_WITH_ESCAPE` is `value$$`, which is substituted without escaping                                                          |
```
@mx-psi here's an example which shows that an environment variable whose value has the escape sequence in it is substituted as is, and not escaped recursively

jack-berg commented on specification/configuration/data-model.md:
```diff
@@ -94,55 +117,48 @@ example, see references to `INVALID_MAP_VALUE` environment variable below.
+| `key: $$$${STRING_VALUE}`                  | `key: $${STRING_VALUE}`             | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, `$$` escape sequence is replaced with `$`, `{STRING_VALUE}` does not match substitution syntax         |
+| `key: $${STRING_VALUE:-fallback}`          | `${STRING_VALUE:-fallback}`         | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, `{STRING_VALUE:-fallback}` does not match substitution syntax                                          |
+| `key: $${STRING_VALUE:-${STRING_VALUE}}`   | `${STRING_VALUE:-value}`            | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, leaving `{STRING_VALUE:-${STRING_VALUE}}`, `${STRING_VALUE}` is replaced with `value`                  |
+| `key: ${UNDEFINED_KEY:-$${UNDEFINED_KEY}}` | `${STRING_VALUE:-${UNDEFINED_KEY}}` | `tag:yaml.org,2002:str`   | `$$` escape sequence is replaced with `$`, leaving `${UNDEFINED_KEY:-` before and `{UNDEFINED_KEY}}` after which do not match substitution syntax |
```
These are interesting examples which show how the escape sequence interacts with the default value. The combination of left to right processing with regex matching on the segments between escape sequences results in behavior which is somewhat surprising on first glance. But it makes sense when you consider the algorithm defined above.

Escaping will always be something users will have to be careful with, especially in the edge cases.
